neps/nep-0491.md

```diff
@@ -0,0 +1,386 @@
+could be returned to the original account that created the account when an
+account is deleted.
+
+The community discussions ended with the conclusion that this feature would
+probably not be used and we should not implement it until there is real demand
+for it.
+
+### Alternative: Don't use smart contracts on user accounts
+
+Instead of deploying contracts on the user account, one could build a similar
+solution that uses zero balance accounts and a single master contract that
+performs all smart contract functionality required. This master contract can
+implement the [Storage Management]
+(https://nomicon.io/Standards/StorageManagement) standard to limit storage usage
+per user.
+
+This solution is not as flexible. The master account cannot make cross-contract
+function calls with the user id as the predecessor.
+
+### Alternative: Move away from storage staking
+
+We could also abandon the concept of storage staking entirely. However, coming
+up with a scalable, sustainable solution that does not suffer from the same
+refund problems is hard.
+
+One proposed design is a combination of zero balance accounts and code sharing
+between contracts. Basically, if somehow the deployed code is stored in a way
+that does not require storage staking by the user themself, maybe the per-user
+state is small enough to fit in the 770 bytes limit of zero balance accounts.
+(Questionable for non-trivial use cases.)
+
+This alternative is much harder to design and implement. The proposal that has
+gotten the furthest so far is [Ephemeral
+Storage](https://github.com/near/NEPs/pull/485), which is pretty complicated and
+does not have community consensus yet. Nobody is currently working on moving it
+forward. While we could wait for that to eventually make progress, in the
+meantime, the community is held back in their innovation because of the refund
+problem.
+
+### Alternative: Using a proxy account
+
+As suggested by [@mfornet](https://github.com/near/NEPs/pull/491#discussion_r1349496234)
+another alternative is using a proxy account approach where the business creates
+a account with deployed contract that has Regular (user has full access key) and
+Restricted mode (user doesn't have full access key and cannot delete account). 
+
+In restricted mode the user has a `FunctionCallKey` function which allows the
+user to call methods of the contract that controls the `FullAccessKey` and 
+allows user some functionality but not all, e.g. not allowing account deletion.
+The user in restricted mode could also upgrade an account by sending the 
+initial amount of NEAR deposited by the account creator and will attach a new
+`FullAccessKey`.
+
+The downside is that accounts in restricted mode do not behave as normal 
+accounts and need to do all actions through `FunctionCall`. To support
+the restricted accounts protocol upgrade is needed, but that can be mitigated by
+adding an alternative key `GranularAccessKey` similar to `FullAccessKey` but
```

Comment:
This idea of a `GranularAccessKey` sounds like a separate alternative from the idea of a proxy account. The proxy account idea does not require a protocol change (the user is only given a FunctionCall access key, which already exists). The complexity with the proxy contract idea is on the tooling side because actions like adding access keys to the account need to be converted to function calls instead of being direct actions.

Another complexity is in developing the contract a business would use. Since there is no protocol-level way to compose functionality from multiple Wasm modules (Account Extensions is on hold), the business would need to include the proxy logic with their business logic in the same contract. Including both functionalities together increases the development complexity of such a contract.